### DISCUSSION TITLE
Is there a way to list installed packages?

### DISCUSSION THREAD

**Discussion by tadayosi** [aquaproj/aqua#2709] [Deprecated (Question and Answer)] [answered]
### Question

`aqua list` only list all the available packages. But sometimes I want to just list what packages I installed via aqua with something like `aqua list --installed`. Looking at other package managers, for examples, brew provides `brew list` and npm provides `npm list` to achieve it.

Is there a way to achieve the same thing with aqua?

**Comment by tadayosi** [comment]
Of course we can do `cat aqua.yaml`, but it's a bit too primitive. It would be better if the cli provides such a feature.

**Reply by suzuki-shunsuke** [reply]
Thank you for your proposal.
Unfortunately, there is no way list all the available packages at the moment.

I think it's good to enhance `aqua list` or `aqua info` commands.

**Comment by suzuki-shunsuke** [comment]
> But sometimes I want to just list what packages I installed via aqua with something like aqua list --installed

May I ask why sometimes you want to just list installed packages?

We don't want to add a feature for just other tools support it.
We would like to clarify use cases and issues the feature would solve.

**Reply by tadayosi** [reply]
Because aqua is a cli manager, I'd like to use it to maintain the packages I installed on my laptop. For maintenance purposes, sometimes I need to know where a cli is installed: dnf, sdkman, aqua, manually, etc.

Another reason I can think up is to automate some operations with aqua in a script. If aqua provided such a feature, some type of operations would've been easier to automate.

**Reply by suzuki-shunsuke** [reply]
> Another reason I can think up is to automate some operations with aqua in a script. If aqua provided such a feature, some type of operations would've been easier to automate.

Could you describe more details?
What do you want to automate?

**Reply by suzuki-shunsuke** [reply]
> I need to know where a cli is installed

How about `aqua which` command?

```sh
aqua which <command>
```

e.g.

```console
$ aqua which go
/Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.22.1.darwin-arm64.tar.gz/go/bin/go
```

Even if the command isn't managed by aqua, `aqua which` returns the path.

```console
$ aqua which ls
/bin/ls
```

I guess this isn't what you want, but I suggest this once as I'd like to understand your needs correctly.

**Reply by tadayosi** [reply]
> How about aqua which command?
> ```
> aqua which <command>
> ```

Sorry, what I meant by "to know where a cli is installed" is that where (= dnf, sdkman, aqua, etc.) I installed the cli tool (e.g. `go`). So `aqua which` doesn't help in my case.

> Could you describe more details?
> What do you want to automate?

This is just an additional idea and I don't have any concrete use cases yet. My primary motivation is the first one so if you don't like the automation idea you can ignore it for now.

**Reply by tadayosi** [reply]
So, to rephrase my use case a bit, sometimes I install for instance `go` with `dnf install golang`, sometimes with `aqua g -i golang/go`, or I might install it manually by downloading it and put it to `~/.local/bin`. Later I might want to uninstall `go`. So the first thing I'd do would be to find out where (= which cli/package manager) I used to install `go`. That's when I need the proposed feature of listing installed packages with aqua.

**Reply by suzuki-shunsuke** [reply]
> Sorry, what I meant by "to know where a cli is installed" is that where (= dnf, sdkman, aqua, etc.) I installed the cli tool (e.g. go). So aqua which doesn't help in my case.

So you want to know how tools such as Go are installed, right?
You're using multiple managers such as dnf, sdkman, aqua, so you want to know which managers you used to install specific tools with, right?

I'm not familiar with tools other than aqua, but I think probably you can understand how tools are installed by checking the output of `aqua which`.

For example, you can understand Go is installed by aqua from the following output.

```console
$ aqua which go
/Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.22.1.darwin-arm64.tar.gz/go/bin/go
```

And you can also understand [ag](https://github.com/ggreer/the_silver_searcher) is installed by Homebrew.

```console
$ aqua which ag
/opt/homebrew/bin/ag
```

This is a small script.

```sh
#!/usr/bin/env bash

set -euo pipefail

cmd=$1

out=$(aqua which "$cmd")
case "$out" in
  *aquaproj-aqua*)
    echo aqua
    ;;
  *homebrew*)
    echo homebrew
    ;;
  *)
    echo "$out"
esac
```

```console
$ bash which-tool.sh ag
homebrew

$ bash which-tool.sh hugo
aqua

$ bash which-tool.sh ls  
/bin/ls
```

> This is just an additional idea and I don't have any concrete use cases yet. My primary motivation is the first one so if you don't like the automation idea you can ignore it for now.

I like automation generally, but if we develop a feature without any concrete use cases, we can't design any sophisticated interface and we may develop a useless feature.

**Reply by tadayosi** [reply]
> So you want to know how tools such as Go are installed, right?

No, there's still miscommunication. Probably because my explanation has been poor :slightly_smiling_face: I'm not so much interested in how each tool installs a package to which location internally. 

> You're using multiple managers such as dnf, sdkman, aqua, so you want to know which managers you used to install specific tools with, right?

Yes, exactly. I just sometimes forget which package manager I used to install a specific cli tool. I feel that using some script like you kindly suggested just for such a case is a bit overkill. I rather just want to run `aqua list --installed` to check a cli/package is installed via aqua. I believe that's the main reason why other managers like dnf, homebrew, npm, nvm, rvm, etc. provide the listing feature.

**Reply by suzuki-shunsuke** [reply]
I see. Thank you for your explanation.

> I rather just want to run aqua list --installed to check a cli/package is installed via aqua.

As I mentioned, you can do this by `aqua which` command.
If the tool is installed via aqua, the output is the path under aqua root directory (`aqua root-dir`), so you can see if the tool is installed via aqua at a glance.

But indeed, if you want to check multiple tools, a command like `aqua list --installed` may be useful.

**Reply by suzuki-shunsuke** [reply]
I think it's easy to implement this feature so I'll try it when I have time.
Thank you for discussing this with me.

**Reply by suzuki-shunsuke** [reply]
- https://github.com/aquaproj/aqua/pull/2733

**Reply by suzuki-shunsuke** [reply]
@tadayosi 

https://github.com/aquaproj/aqua/releases/tag/v2.24.0-1

Could you try v2.24.0-1 and give us your feedback?

```console
$ aqua upa v2.24.0-1
$ aqua list -installed [-a]
```

e.g.

```console
$ aqua list -installed
standard,rhysd/actionlint,v1.6.27
standard,suzuki-shunsuke/cmdx,v1.7.4
standard,sigstore/cosign,v1.13.2
standard,suzuki-shunsuke/ghalint,v0.2.9
standard,int128/ghcp,v1.13.2
standard,golangci/golangci-lint,v1.56.2
standard,goreleaser/goreleaser,v1.24.0
standard,reviewdog/reviewdog,v0.17.1
```

**Reply by suzuki-shunsuke** [reply]
https://github.com/aquaproj/aqua/releases/tag/v2.24.0-2

Changed the order and separator to make the output more human friendly.

e.g.

```console
$ aqua list -installed   
rhysd/actionlint	v1.6.27	standard
suzuki-shunsuke/cmdx	v1.7.4	standard
sigstore/cosign	v1.13.2	standard
suzuki-shunsuke/ghalint	v0.2.9	standard
int128/ghcp	v1.13.2	standard
golangci/golangci-lint	v1.56.2	standard
goreleaser/goreleaser	v1.24.0	standard
reviewdog/reviewdog	v0.17.1	standard
```

**Reply by tadayosi** [reply]
@suzuki-shunsuke Thank you very much for such a quick release!  It works great to me 😍 !

**Reply by suzuki-shunsuke** [reply]
Thank you for your feedback!

- https://github.com/aquaproj/aquaproj.github.io/pull/838

**Reply by suzuki-shunsuke** [reply, marked as answer]
v2.24.0 is out 🎉 

https://github.com/aquaproj/aqua/releases/tag/v2.24.0
